homeassistant/components/tesla_fleet/config_flow.py

```diff
@@ -27,6 +31,23 @@ def logger(self) -> logging.Logger:
         """Return logger."""
         return LOGGER
 
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle a flow start."""
+        implementations = await config_entry_oauth2_flow.async_get_implementations(
+            self.hass, self.DOMAIN
+        )
+        if not implementations:
```

Comment:
This isn't valid, as multiple can be set up. You'll never know, at this point, which credentials have been used.

This may result an in existing config entry break when a new application credential is created.

../Frenck

Reply:
So I've been testing this, and as far as I can tell it's safe to create the credential when one doesn't exist.

However you're suggesting it's not?

Is there a way to creating an application credential for the user then?

Reply:
> and as far as I can tell it's safe to create the credential when one doesn't exist.

Yup that is safe, my point it the other way around.

If there is already a config entry using the one you add below. If a application credential is added, the one blow will not be added anymore after a restart, meaning the existing config entry will be pointing to a application credential that isn't added anymore.

Reply:
I think we need to check if the OSS credentials are registered or not.

Reply:
> the one below will not be added anymore after a restart

But application credentials persist restarts, so the one I created below will still be there. I suspect I'm misunderstanding something about this.

At a high level, the user should only ever need to use the single application credential that gets created for them.

The integration actually doesn't support the user creating their own application credential at this time, and will completely ignore their inputs if they do (but still work). I have a PR ready to add custom credential support in 2024.9

I don't think I understand the failure scenario you're describing

Reply:
Hmmm I see your point, and that makes me see what I missed earlier.

This built-in credential should not be imported, but registered instead.

Reply:
So ive implemented `async_register_implementation` by copying ondilo_ico, but I'll be honest, I dont fully understand how it works, but it does appear to be fully working.

`async_register_implementation` doesnt create application credentials right?

Reply:
>  doesnt create application credentials right?

It does, it registers one provided by the system (not by the user).